ntoskrnl/se/token.c

```diff
@@ -2220,78 +2244,83 @@ SeTokenCanImpersonate(
 
-    /*
-     * SecurityAnonymous and SecurityIdentification levels do not
-     * allow impersonation.
-     */
-    if (ImpersonationLevel == SecurityAnonymous ||
-        ImpersonationLevel == SecurityIdentification)
-    {
-        return FALSE;
-    }
-
     /* Time to lock our tokens */
     SepAcquireTokenLockShared(ProcessToken);
     SepAcquireTokenLockShared(TokenToImpersonate);
 
-    /* What kind of authentication ID does the token have? */
-    if (RtlEqualLuid(&TokenToImpersonate->AuthenticationId,
-                     &SeAnonymousAuthenticationId))
+    /* Allow impersonation for the process server if it's granted the impersonation privilege */
+    if ((ProcessToken->TokenFlags & TOKEN_HAS_IMPERSONATE_PRIVILEGE) != 0)
     {
-        /*
-         * OK, it looks like the token has an anonymous
-         * authentication. Is that token created by the system?
-         */
-        if (TokenToImpersonate->TokenSource.SourceName != SeSystemTokenSource.SourceName &&
-            !RtlEqualLuid(&TokenToImpersonate->TokenSource.SourceIdentifier, &SeSystemTokenSource.SourceIdentifier))
-        {
-            /* It isn't, we can't impersonate regular tokens */
-            DPRINT("SeTokenCanImpersonate(): Token has an anonymous authentication ID, can't impersonate!\n");
-            CanImpersonate = FALSE;
-            goto Quit;
-        }
+        DPRINT("The process is granted the impersonation privilege, allow impersonation\n");
+        SepReleaseTokenLock(TokenToImpersonate);
+        SepReleaseTokenLock(ProcessToken);
+        return TRUE;
     }
 
-    /* Are the SID values from both tokens equal? */
-    if (!RtlEqualSid(ProcessToken->UserAndGroups->Sid,
-                     TokenToImpersonate->UserAndGroups->Sid))
+    /*
+     * As the name implies, an anonymous token has invisible security
+     * identification details. By the general rule these tokens do not
+     * pose a danger in terms of power escalation so give the server a pass.
+     */
+    if (RtlEqualLuid(&TokenToImpersonate->AuthenticationId,
+                     &SeAnonymousAuthenticationId))
     {
-        /* They aren't, bail out */
-        DPRINT("SeTokenCanImpersonate(): Tokens SIDs are not equal!\n");
-        CanImpersonate = FALSE;
-        goto Quit;
+        DPRINT("The token to impersonate has an anonymous authentication ID, allow impersonation either way\n");
+        SepReleaseTokenLock(TokenToImpersonate);
+        SepReleaseTokenLock(ProcessToken);
+        return TRUE;
     }
 
     /*
-     * Make sure the tokens aren't diverged in terms of
-     * restrictions, that is, one token is restricted
-     * but the other one isn't.
+     * Deny impersonation for the server if it wants to impersonate a client
+     * beyond what the impersonation level originally permits.
      */
-    if (SeTokenIsRestricted(ProcessToken) !=
-        SeTokenIsRestricted(TokenToImpersonate))
+    if (ImpersonationLevel > TokenToImpersonate->ImpersonationLevel)
```

Comment:
Is this what Windows does?

Reply:
No, nor does it mean that it is automatically disqualified because it is not what Windows does. Having the check there makes sense because you want a server that respects the impersonation limitations imposed by a client, e.g. if the client forbids anyone to delegate its security credentials through remote impersonation but only impersonate the client, you can't simply ask to impersonate other than what you've been limited to.